models/classes/runner/time/storageFormat/QtiTimeStoragePackedFormat.php

```diff
@@ -273,24 +274,42 @@ public function decode($data)
                 }
 
                 foreach ($decodedData as $key => &$value) {
-                    if (is_array($value)) {
+                    if ($key === self::STORAGE_KEY_TIME_LINE) {
                         $decodedData[$key] = $this->unpackTimeLine($value);
                     }
+                    if ($key === self::STORAGE_KEY_TIMER_ADJUSTMENT_MAP) {
+                        $decodedData[$key] = $this->decodeAdjustmentMap($value);
+                    }
                 }
 
                 unset($decodedData[self::STORAGE_KEY_FORMAT]);
                 unset($decodedData[self::STORAGE_KEY_VERSION]);
             } else {
                 foreach ($decodedData as $key => &$value) {
-                    if (is_array($value)) {
+                    if ($key === self::STORAGE_KEY_TIME_LINE && !$value instanceof QtiTimeLine) {
                         $timeLine = new QtiTimeLine();
                         $timeLine->fromArray($value);
                         $decodedData[$key] = $timeLine;
                     }
+                    if ($key === self::STORAGE_KEY_TIMER_ADJUSTMENT_MAP && !$value instanceof AdjustmentMap) {
+                        $decodedData[$key] = $this->decodeAdjustmentMap($value);
+                    }
                 }
             }
         }
 
         return $decodedData;
     }
+
+    /**
+     * @param $value
+     * @return AdjustmentMap
+     */
+    private function decodeAdjustmentMap($value)
+    {
+        $map = new AdjustmentMap();
+        $map->fromArray($value);
```

Comment:
could it be a static generator? $this->getServiceLocator(AdjustmentMap::class)->getMap($value) // or createMap or generateMap?

Reply:
I think it is more consistent when all classes that are stored in `TimeStorage` use the same interface for serialization. `QtiTimeLine` is using `oat\taoTests\models\runner\time\ArraySerializable`.